Getting Started with JavaScript (#12689)
WordPress/gutenberg: docs/designers-developers/developers/tutorials/javascript/readme.md

# Getting Started with JavaScript

> The purpose of this tutorial is to step through getting started with JavaScript and WordPress. The Gutenberg documentation contains information on the APIs available, but as a reference and not necessarily full working examples. The goal of this tutorial is get you comfortable on how to use these snippets of code.

Reviewer: Would it be worthwhile to have a short explanation of "What is JavaScript"?

Inside of this new directory, create a file called `myguten-plugin.php` which will be the server-side code the runs when your plugin is active. For now you can place the following in that file:

```php
<?php
/*
Plugin Name: My Guten Plugin
*/
```

So you should have a directory `wp-content/plugins/myguten-plugin/` which has the single file `myguten-plugin.php`. Once that is in place, you can go to your plugins list in `wp-admin` and you should see your plugin listed.

Click **Activate** and your plugin will load with WordPress.

## Loading JavaScript

Now with the plugin in place, we can add our code that loads the JavaScript we want to use. This follows the standard WordPress methodology of enqueuing your scripts, see [Including CSS & JavaScript](https://developer.wordpress.org/themes/basics/including-css-javascript/) for more details.

Add the following code to your `myguten-plugin.php` file:

```php
function myguten_enqueue( $hook ) {
	wp_enqueue_script( 'myguten-script',
		plugins_url( 'myguten.js', __FILE__ ),

> 		array( 'wp-blocks', 'wp-element' )

Reviewer: Is it worth having a code comment explaining what each of these function arguments is? I think the note you have below is less obvious.

Author: Rearranged and updated information on dependencies.

	);
}
add_action( 'admin_enqueue_scripts', 'myguten_enqueue' );
```

Create a file called `myguten.js` and add:

```js
console.log( "I'm loaded!" );

> ```

Reviewer: Probably worth recapping each of the files you have at this point.

You could also offer a direct download of this basic stub plugin to help people get going.

Reviewer 2: I am wondering if we should use the `gutenberg-examples` repo for storing code associated with specific tutorials like this?

Author: The tutorial was updated and a recap added.

I like the idea of a Github repo to put the code, but the `gutenberg-examples` is so tailored for it being a plugin, it might not be the right fit. A `gutenberg-tutorials` ?  We can discuss further in #docs maybe even move content out of handbook and link to make them more stand-alone.

Now go to any page within `/wp-admin/` and check your browser Developer Tools, and you should see the message in your console log. See [Mozilla's What are browser developer tools?](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/What_are_browser_developer_tools) if you need more information, the area to become most familiar with is [The JavaScript console](https://developer.mozilla.org/en-US/docs/Learn/Common_questions/What_are_browser_developer_tools#The_JavaScript_console).

> The script enqueuing used the `admin_enqueue_scripts` hook which only loads the JavaScript within the wp-admin section. So if you navigate to any front-end page or article on your site, you will not see the message.

Reviewer: I don't think `admin_enqueue_scripts` is the correct action for enqueuing block assets...

Author: You're right, updated. I was reading through the core plugin/theme documentation and trying to use those examples the user might be familiar with to match the same ideas to working with Gutenberg.

**Note:** The last argument in the `wp_enqueue_script()` function is an array of dependencies, all of the Gutenberg packages are registered and can be loaded by specifying them in the array, blocks and elements are shown as two common examples.

Recap at this point, we have a plugin which loads JavaScript, we're off to a good start.

## JavaScript versions and building

> The Gutenberg Handbook shows examples in two syntaxes ES5 and ESNext, note ESNext is sometimes referred to as ES6. The ES5 code is compatible with almost all current browsers.

Reviewer: ```suggestion
The Gutenberg Handbook shows examples in two variations of JavaScript, ES5 and ESNext. What's even more confusing is that ESNext is sometimes referred to as ES6. The ES5 code is compatible with almost all current browsers.
```

The ESNext syntax is compatible with most modern browsers, but unfortunately not IE11. Additionally, the ESNext code examples include JSX syntax, which requires a build step using webpack to transform into code that can be run in your browser.

> For this tutorial, all examples will be in straight ES5 code that can be run straight in your browser and does not require an additional build step.

Reviewer: ```suggestion
For this tutorial, all examples are written as the ES5 variation of JavaScript that can be run straight in your browser and does not require an additional build step.
```

## Extending the Block Editor

This puts all the initial pieces in place for you to start extending the Block Editor.

Let's look at using the [Block Style Variation example](../../../../../docs/designers-developers/developers/filters/block-filters/#block-style-variations).

> Update the file `myguten.js` with:

Reviewer: ```suggestion
Replace the existing `console.log()` code in your `myguten.js` with:
```

```js
wp.blocks.registerBlockStyle( 'core/quote', {
    name: 'fancy-quote',
    label: 'Fancy Quote'
} );
```

> You can remove the previous `console.log` if you like.

Reviewer: ```suggestion

```

Author: Ah yes, "Replace". Thanks, I found the phrasing awkward when I wrote it but a better way didn't come to mind

> After you add the code, save the file, and start a new post in the Block Editor.

Reviewer: ```suggestion
After you add the `wp.blocks.registerBlockStyle` code, save the `myguten.js` file, and then create a new post in the Block Editor.
```

> Add a quote block, and in the right sidebar under Styles, you should see your new Fancy Quote style listed. You can go back to the JavaScript and change the label to "Fancy Pants" and reload, and you should see the new label.

Reviewer: Worth including a screenshot here?

Previewing or Publishing the post, you will not see a visible change. However, if you look at the source, you should see `is-style-fancy-quote` class name attached to your quote.

> You could create a `style.css` file with:

Reviewer: ```suggestion
Go ahead and create a `style.css` file with:
```

```css
.is-style-fancy-quote {
	font-size: 64px;
}

```

and enqueue the CSS by adding the following to your `myguten-plugin.php`:

```php

> function myguten_stylesheet() {

Reviewer: Can this function have PHPDoc too, explaining what it's doing?

Author: I've broken up the code and changed things around with more explanations above the code. My concern is if it were as a comment in code people will miss it. I've also included links to all referred to functions for any additional details are available.

	wp_enqueue_style( 'myguten-style', plugins_url( 'style.css', __FILE__ ) );
}

> add_action( 'wp_enqueue_scripts', 'myguten_stylesheet' );

Reviewer: ```suggestion
add_action( 'enqueue_block_assets', 'myguten_stylesheet' );
```

```

> And then when you view the page, you should see it in a very large font.

Reviewer: ```suggestion
And then when you view the page, you will see it in a very large font.
```

> 

Reviewer: Would it be worthwhile to include some "But it's not working?" debugging steps?

Author: 👍
